backends/p4tools/testgen/core/small_step/table_stepper.h

```diff
@@ -0,0 +1,207 @@
+
+        /// Whether the table is constant and can not accept control plane entries.
+        bool tableIsImmutable = false;
+
+        /// Ordered list of key fields with useful properties.
+        std::vector<KeyProperties> resolvedKeys;
+    } properties;
+
+ public:
+    /* =========================================================================================
+     * Table Variable Getter functions
+     * ========================================================================================= */
+    /// @returns the boolean-typed state variable that tracks whether given table is reached.
+    static const StateVariable& getTableReachedVar(const IR::P4Table* table);
+
+    /// @returns the state variable that tracks the value read by the key at the given index in the
+    /// given the table.
+    static const StateVariable& getTableKeyReadVar(const IR::P4Table* table, int keyIdx);
+
+    /// @returns the boolean-typed state variable that tracks whether the given table was hit. The
+    /// value of this variable is false if the table misses or is not reached.
+    static const StateVariable& getTableHitVar(const IR::P4Table* table);
+
+    /// @returns the state variable that tracks the index of the action taken by the given table.
+    ///
+    /// This variable is initially set to the number of actions in the table, indicating that no
+    /// action has been selected. It is set by setTableAction and read by getTableAction.
+    static const StateVariable& getTableActionVar(const IR::P4Table* table);
+
+ protected:
+    /* =========================================================================================
+     * Table Utility functions
+     * ========================================================================================= */
+    /// @returns the execution state of the friend class ExpressionStepper.
+    const ExecutionState* getExecutionState();
+
+    /// @returns the program info of the friend class ExpressionStepper.
+    const ProgramInfo* getProgramInfo();
+
+    /// @returns the result pointer of the friend class ExpressionStepper.
+    ExprStepper::Result getResult();
+
+    /// The function compares the right and left expression.
+    /// Returns 0 if the values are equal, 1 if the left value is greater,
+    /// and -1 if the right value is greater.
+    /// If the left value is constant entries and the right value
+    /// is Mask/Default then priority is given to constants.
+    /// For more detailed information see
+    /// https://github.com/p4lang/behavioral-model/blob/main/docs/simple_switch.md#longest-prefix-match-tables
+    int compareLPMEntries(const IR::Expression* left, const IR::Expression* right);
+
+    /// Sorts entries for lpm.
+    /// TODO: Check whether we still need this
+    IR::Vector<IR::Entry> sortLpmEntries(const IR::Vector<IR::Entry>& entries);
+
+    /// @returns whether the table is immutable, meaning control plane entries can not be added.
```

Comment:
the default_action can also be const

Reply:
But that just means that we can not override the default action, right?

Reply:
yes, the same as for the entries

Reply:
I believe that whether the entries are const (i.e. not changeable by the control plane), and whether the default_action is const (i.e. not changeable by the control plane) are independent properties of a P4 table.  All four combinations of (const entries, const default_action), (const entries, non-const default_action), (non-const entries, const default_action), and (non-const entries, non-const default_action) are supported by p4c and the language spec.

When I say "non-const entries" here, I simply mean that there is no 'entries' property of the table at all in the P4 source code.  Although for future reference, there are people who are encouraging me to help define "entries = { ... }" without `const` for P4 soon-ish.  There is no such language spec definition or p4c implementation for that today that I am aware of.